src/webots/vrml/WbProtoManager.cpp

```diff
@@ -105,85 +107,116 @@ void WbProtoManager::readModel(WbTokenizer *tokenizer, const QString &worldPath)
-      url = QDir::cleanPath(WbStandardPaths::webotsHomePath() + url.mid(9));
+    else {
+      const QString errorMessage =
+        tr("PROTO declarations are missing. Please adapt your project to R2022b "
+           "following these instructions: "
+           "https://github.com/cyberbotics/webots/wiki/How-to-adapt-your-world-or-PROTO-to-Webots-R2022b");
+      if (!mUniqueErrorMessages.contains(errorMessage)) {
+        mUniqueErrorMessages << errorMessage;
+        WbLog::error(errorMessage);
+      }
+    }
+  }
+
+  // a PROTO declaration is provided, enforce it
+  QString modelPath;  // how the PROTO is referenced
+  if (WbUrl::isWeb(protoDeclaration)) {
+    modelPath = protoDeclaration;
+    assert(WbNetwork::instance()->isCached(modelPath));
+  } else if (WbUrl::isLocalUrl(protoDeclaration)) {
+    // two possibitilies arise if the declaration is local (webots://)
+    // 1. the parent PROTO is in the cache (all its references are always 'webots://'): it may happen if a PROTO references
+    // another PROTO (both being cached)
+    // 2. the PROTO is actually locally available
+    // option (1) needs to be checked first, otherwise in the webots development environment the declarations aren't
+    // respected (since a local version of the PROTO exists virtually every time)
+    QString parentFile = parentFilePath;
+    if (parentFile.startsWith(WbNetwork::instance()->cacheDirectory()))
+      // reverse lookup the file in order to establish its original remote path
+      parentFile = WbNetwork::instance()->getUrlFromEphemeralCache(parentFile);
+
+    // extract the prefix from the parent so that we can build the child's path accordingly
+    if (WbUrl::isWeb(parentFile)) {
+      QRegularExpression re("(https://raw.githubusercontent.com/cyberbotics/webots/[a-zA-Z0-9\\_\\-\\+]+/)");
+      QRegularExpressionMatch match = re.match(parentFile);
+      if (match.hasMatch()) {
+        // inject the prefix based on that of the parent
+        modelPath = protoDeclaration.replace("webots://", match.captured(0));
+        // if the PROTO tree was built correctly, by definition the child must be cached already too
+        assert(WbNetwork::instance()->isCached(modelPath));
+      } else {
+        WbLog::error(tr("The cascaded url inferring mechanism is supported only for official webots assets."));
+        return NULL;
+      }
+    } else
+      modelPath = WbUrl::computePath(protoDeclaration);
+  } else
+    modelPath = WbUrl::computePath(protoDeclaration);
+
+  // determine prefix and disk location from modelPath
+  const QString modelDiskPath = WbUrl::isWeb(modelPath) ? WbNetwork::instance()->get(modelPath) : modelPath;
+  const QString prefix = WbUrl::computePrefix(modelPath);  // used to retrieve remote assets (replaces webots:// in the body)
 
-    WbProtoModel *model = readModel(QFileInfo(url).absoluteFilePath(), worldPath, url, baseTypeList);
+  if (QFileInfo(modelDiskPath).exists() && !modelPath.isEmpty()) {
```

Comment:
I would switch the conditions and list first the quicker to check to 
```suggestion
  if (!modelPath.isEmpty() && QFileInfo(modelDiskPath).exists()) {
```